src/Conversion/ONNXToKrnl/Math/Elementwise.cpp

```diff
@@ -1621,6 +1729,20 @@ struct ONNXElementwiseVariadicOpLowering
             alignment, simdUnroll);
     }
 
+    // Try to fuse the unary elementwise consumers
+    SmallVector<Operation *, 2> fusibleOps;
+    SmallVector<EmitScalarFunc, 2> fuseEmitFunctions;
+    bool fusibleOpsFound =
+        findFusibleUnaryOps(op, fusibleOps, fuseEmitFunctions);
+    if (fusibleOpsFound) {
+      // After fusion, the only store is for the last Op.
+      // Therefore, the allocation should be the output of the last Op
+      Operation *lastOp = fusibleOps[fusibleOps.size() - 1];
+      outputMemRefType =
+          this->typeConverter->convertType(lastOp->getResults()[0].getType())
+              .template cast<MemRefType>();
+    }
```

Comment:
This looks redundant. The type of the last op would be similar to the type of the current op, since all fused ops are unary ops. Otherwise, we cannot use shapeHelper of the current op in the `alignedAlloc`. In other words, it would be fine to alloc the output buffer by using the current op's output type.

Reply:
I add a test case for dynamic size with broadcast. The alignedAlloc takes the shapeHelper of the current Op for the dynamic dimensions and the memory ref for the type. Since the element-wise ops do not change the shape of. tensor. alignedAlloc works. We cannot use the shapeHelper of the last Op because the inputs are not available for shape propagation.

Reply:
Got it. Thanks!

Reply:
I really ran into an example of cast in the backend test. For test_dynamicquantizelinear_expanded.onnx, part of the list of ops are listed below:
```
   %9 = "onnx.Sub"(%0, %8) : (tensor<f32>, tensor<f32>) -> tensor<f32> 
   %10 = "onnx.Clip"(%9, %0, %1) : (tensor<f32>, tensor<f32>, tensor<f32>) -> tensor<f32>
    %11 = "onnx.Round"(%10) : (tensor<f32>) -> tensor<f32>
    %12 = "onnx.Cast"(%11) {to = ui8} : (tensor<f32>) -> tensor<ui8>
    %13 = "onnx.QuantizeLinear"(%arg0, %7, %12) {axis = 1 : si64} : (tensor<6xf32>, tensor<f32>, tensor<ui8>) -> tensor<6xui8>
    return %13, %7, %12 : tensor<6xui8>, tensor<f32>, tensor<ui8>
```
Clip, Round, and Cast  are fused. I fixed an error for this test case: the Clip should not be fused with Sub. Though Clip is lowered as ElementwiseUnaryOp in our implementation, it has three inputs. I guess the two optional inputs are added in the newer version of onnx. The root of the problem came from the interface: we pass a ValueRange for input but did not specify the expected size for unary and binary Op. The ValueRange is for variadic Ops.
I guess @tungld and @AlexandreEichenberger may be interested in this issue.

Reply:
Yes, we treat Clip as a unary op since the last two tensors are scalar. We had been passing the two tensors as they are to emitScalarOpFor<ONNXClipOp>, but recently their loaded values are passed to instead that would be convenient for SIMD I guess.